### PR TITLE
dedicated field for cluster_label

### LINKED ISSUE #6158: Cluster.label breaking if alertmanagers have different names
**What happened?**
We recently upgraded the alertmanager from v0.23.0 to v0.26.0 and our cluster stopped settling gossip (any version below v0.26.0 should have the problem).
The cluster label was set differently on each alertmanager, since they have different names.
These were the arguments the operator generated and attached to AM:
```
--cluster.label=prometheus-alertmanager/am-az-1
--cluster.label=prometheus-alertmanager/am-az-2
```

**Did you expect to see something different?**
It wasn't explicitly mentioned as breaking so a non communicating cluster was not expected.

**How to reproduce it (as minimally and precisely as possible)**:
Take two dedicated clusters, deploy an alertmanager on each, cluster them and name them differently:
* am-az-1
* am-az-2

**Environment**

* Prometheus Operator version: `0.70.0 / 2023-11-30`

* Kubernetes version information: `v1.28.2`

**Anything else we need to know?**:

The code itself implies that this might be an issue, apparently no one has run into it yet - or never reported it. Normally the names of the alertmanagers would be the same across clusters I suppose, however there might be others with the same issue.

Would you be willing to take a PR with the respective change to configure `cluster.label`?

### PR DESCRIPTION
**Pull request by viennaa**
## Description

Fixes #6158.

It adds a dedicated field for cluster.label option in alertmanager. Default is to leave it blank, if it is not set. More background is mentioned in #6158 

## Type of change

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [x] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

```release-note
alertmanager `--cluster.label` (introduced `v0.26.0`) can be reconfigured with a dedicated field now.
```

### CODE REVIEW

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 1]
```diff
@@ -206,6 +206,8 @@ type AlertmanagerSpec struct {
 	ClusterAdvertiseAddress string `json:"clusterAdvertiseAddress,omitempty"`
 	// Interval between gossip attempts.
 	ClusterGossipInterval GoDuration `json:"clusterGossipInterval,omitempty"`
+	// Gossip with same ClusterLabels only.
+	ClusterLabel string `json:"clusterLabel,lomitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:210]:
```suggestion
	ClusterLabel *string `json:"clusterLabel,omitempty"`
```

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 2]
```diff
@@ -206,6 +206,8 @@ type AlertmanagerSpec struct {
 	ClusterAdvertiseAddress string `json:"clusterAdvertiseAddress,omitempty"`
 	// Interval between gossip attempts.
 	ClusterGossipInterval GoDuration `json:"clusterGossipInterval,omitempty"`
+	// Gossip with same ClusterLabels only.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:209]:
```suggestion
	// Defines the identifier that uniquely identifies the Alertmanager cluster.
	// You should only set it when the Alertmanager cluster includes several Alertmanager objects.
```

File: pkg/alertmanager/statefulset.go [hunk 3]
```diff
@@ -313,10 +313,11 @@ func makeStatefulSetSpec(logger log.Logger, a *monitoringv1.Alertmanager, config
...
 	// --cluster.label flag was introduced in alertmanager v0.26, this helps to block
 	// any traffic that is not meant for the cluster.
-	// The value is hardcoded and the value is guaranteed to be unique in a given cluster but
-	// if there's a use case, we can consider a new field in the CRD.
+	// Specify ClusterLabel in case you are in need of sorting out Alertmanager traffic.
 	if version.GTE(semver.MustParse("0.26.0")) {
-		amArgs = append(amArgs, fmt.Sprintf("--cluster.label=%s/%s", a.Namespace, a.Name))
+		if a.Spec.ClusterLabel != "" {
+			amArgs = append(amArgs, fmt.Sprintf("--cluster.label=%s", a.Spec.ClusterLabel))
+		}
```
Comment (simonpasquier) [inline pkg/alertmanager/statefulset.go:320]:
```suggestion
		clusterLabel := fmt.Sprintf("%s/%s", a.Namespace, a.Name)
		if a.Spec.ClusterLabel != nil {
			clusterLabel = *a.Spec.ClusterLabel
		}
		amArgs = append(amArgs, fmt.Sprintf("--cluster.label=%s", clusterLabel))
```

**Review by simonpasquier — COMMENTED** [review]
Thanks for the PR!

**Comment by viennaa** [comment]
Pushed the newly generated files and test. 

`checks / Generate and format (ubuntu-latest) (pull_request)` are still failing, don't really get why.

**Comment by ArthurSens** [comment]
> Pushed the newly generated files and test.
> 
> `checks / Generate and format (ubuntu-latest) (pull_request)` are still failing, don't really get why.

Making changes to the API requires us to regenerate a bunch of files, like API documentation, CRD yaml/json files, etc. You can generate them by running the command `make generate --always-make` and commit the changed files :)

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 4]
```diff
@@ -206,6 +206,9 @@ type AlertmanagerSpec struct {
 	ClusterAdvertiseAddress string `json:"clusterAdvertiseAddress,omitempty"`
 	// Interval between gossip attempts.
 	ClusterGossipInterval GoDuration `json:"clusterGossipInterval,omitempty"`
+	// Defines the identifier that uniquely identifies the Alertmanager cluster.
+	// You should only set it when the Alertmanager cluster includes several Alertmanager objects.
```
Comment (ArthurSens) [inline pkg/apis/monitoring/v1/alertmanager_types.go:210]:
```suggestion
	// You should only set it when the Alertmanager cluster includes more than one Alertmanager replicas.
```

File: pkg/alertmanager/statefulset_test.go [hunk 5]
```diff
@@ -1216,11 +1222,19 @@ func TestClusterLabel(t *testing.T) {
...
+				a.Spec.ClusterLabel = &ts.customClusterLabel
+			}
+
 			ss, err := makeStatefulSetSpec(nil, &a, defaultTestConfig, nil)
 			require.NoError(t, err)
 
 			args := ss.Template.Spec.Containers[0].Args
 			if ts.expectedClusterLabelArg {
+				if ts.customClusterLabel != "" {
+					require.Contains(t, args, "--cluster.label=custom.cluster")
```
Comment (ArthurSens) [inline pkg/alertmanager/statefulset_test.go:1235]:
```suggestion
					require.Contains(t, args, fmt.Sprintf("--cluster.label=%s", ts.customClusterLabel))
```

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 6]
```diff
@@ -206,6 +206,9 @@ type AlertmanagerSpec struct {
 	ClusterAdvertiseAddress string `json:"clusterAdvertiseAddress,omitempty"`
 	// Interval between gossip attempts.
 	ClusterGossipInterval GoDuration `json:"clusterGossipInterval,omitempty"`
+	// Defines the identifier that uniquely identifies the Alertmanager cluster.
+	// You should only set it when the Alertmanager cluster includes more than one Alertmanager replicas.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:210]: That's not entirely exact. Even with multiple Alertmanager pods, you shouldn't need to set this field as long as you don't join pods across different clusters.

File: pkg/alertmanager/statefulset.go [hunk 7]
```diff
@@ -313,10 +313,13 @@ func makeStatefulSetSpec(logger log.Logger, a *monitoringv1.Alertmanager, config
 	// connects with cluster A.
 	// --cluster.label flag was introduced in alertmanager v0.26, this helps to block
 	// any traffic that is not meant for the cluster.
-	// The value is hardcoded and the value is guaranteed to be unique in a given cluster but
-	// if there's a use case, we can consider a new field in the CRD.
+	// Specify ClusterLabel in case you are in need of sorting out Alertmanager traffic.
```
Comment (simonpasquier) [inline pkg/alertmanager/statefulset.go:316]: IMHO the field's description should be enough to understand.
```suggestion
```

File: pkg/alertmanager/statefulset_test.go [hunk 8]
```diff
@@ -1193,10 +1193,16 @@ func TestClusterLabel(t *testing.T) {
 		scenario                string
 		version                 string
 		expectedClusterLabelArg bool
+		customClusterLabel      string
 	}{{
 		scenario:                "--cluster.label set by default for version >= v0.26.0",
 		version:                 "0.26.0",
 		expectedClusterLabelArg: true,
+	}, {
+		scenario:                "--cluster.label set if speficfied explicity",
```
Comment (simonpasquier) [inline pkg/alertmanager/statefulset_test.go:1202]:
```suggestion
		scenario:                "--cluster.label set if specified explicitly",
```

File: pkg/apis/monitoring/v1/alertmanager_types.go (hunk 6 above)
Comment (viennaa) [reply pkg/apis/monitoring/v1/alertmanager_types.go:210]: @ArthurSens this was your suggestion, maybe you can chime in?

File: pkg/alertmanager/statefulset_test.go (hunk 8 above)
Comment (viennaa) [reply pkg/alertmanager/statefulset_test.go:1202]: good catch

File: pkg/apis/monitoring/v1/alertmanager_types.go (hunk 6 above)
Comment (simonpasquier) [reply pkg/apis/monitoring/v1/alertmanager_types.go:210]: sorry my initial suggestion wasn't exactly what I had in mind. Refining here.
```suggestion
	// You should only set it when the Alertmanager cluster includes Alertmanager instances which are external to this Alertmanager resource. In practice, the addressed of the external instances are provided via the `.spec.additionalPeers` field.
```

Comment (viennaa) [reply pkg/apis/monitoring/v1/alertmanager_types.go:210]: This describes it way better, you should not worry about it, if you don't have external AMs joining the cluster. Wasn't sure about how extensive this description should be here.
Would take it this way and re-render the package.

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 9]
```diff
@@ -206,6 +206,9 @@ type AlertmanagerSpec struct {
 	ClusterAdvertiseAddress string `json:"clusterAdvertiseAddress,omitempty"`
 	// Interval between gossip attempts.
 	ClusterGossipInterval GoDuration `json:"clusterGossipInterval,omitempty"`
+	// Defines the identifier that uniquely identifies the Alertmanager cluster.
+	// You should only set it when the Alertmanager cluster includes Alertmanager instances which are external to this Alertmanager resource. In practice, the addressed of the external instances are provided via the `.spec.additionalPeers` field.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:210]: (nit)
```suggestion
	// You should only set it when the Alertmanager cluster includes Alertmanager instances which are external to this Alertmanager resource. In practice, the addresses of the external instances are provided via the `.spec.additionalPeers` field.
```

File: pkg/apis/monitoring/v1/alertmanager_types.go (hunk 6 above)
Comment (ArthurSens) [reply pkg/apis/monitoring/v1/alertmanager_types.go]: > @ArthurSens this was your suggestion, maybe you can chime in?

Yeah, sorry for the shallow review, I did not look in depth at the problem. What Simon suggested looks a lot more precise

**Comment by viennaa** [comment]
@simonpasquier are we good for an approval? Let me know if something needs to be done from my side.

**Review by ArthurSens — APPROVED** [review]
LGTM

**Review by simonpasquier — APPROVED** [review]
Thanks!

**Pull request merged**